File: lintcode/python/0074_first_bad_version.py
#!/usr/bin/python3 -t


# binary search
# time O(logn)
# space O(1)

#class SVNRepo:
#    @classmethod
#    def isBadVersion(cls, id)
#        # Run unit tests to check whether verison `id` is a bad version
#        # return true if unit tests passed else false.
# You can use SVNRepo.isBadVersion(10) to check whether version 10 is a
# bad version.
class Solution:
    """
    @param n: An integer
    @return: An integer which is the first bad version.
    """
    def findFirstBadVersion(self, n):
        # write your code here
        start = 1
        end = n
        while start + 1 < end:
            mid = start + (end-start)//2
            if SVNRepo.isBadVersion(mid):
                end = mid
            else:
                start = mid

        if SVNRepo.isBadVersion(start):
            return start
        return end

# binary search solution

#class SVNRepo:
#    @classmethod
#    def isBadVersion(cls, id)
#        # Run unit tests to check whether verison `id` is a bad version
#        # return true if unit tests passed else false.
# You can use SVNRepo.isBadVersion(10) to check whether version 10 is a 
# bad version.
class Solution:
    """
    @param n: An integer
    @return: An integer which is the first bad version.
    """
    def findFirstBadVersion(self, n):
        # write your code here
        l = 1
        r = n
        
        while l < r:
            mid = (l+r)/2
            if SVNRepo.isBadVersion(mid):
                r = mid
            else:
                l = mid+1
                
        return l

File: lintcode/golang/0074_first_bad_version.go
/**
 * @param n: An integer
 * @return: An integer which is the first bad version.
 */
func findFirstBadVersion (n int) int {
    // write your code here
    if n == 0 {
        return 0
    }
    
    l : = 0
    r := n-1
    
    for l + 1 < r {
        mid = (l+r)/2
        if isBadVErsion(mid) {
            r = mid
        } else {
            l = mid
        }
    }
    
    if isBadVErsion(l) {
        return l
    }
    
    return r
}

